// binarytf/dist/lib/util/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var BinaryTokens;
(function (BinaryTokens) {
    BinaryTokens[BinaryTokens["NullPointer"] = 0] = "NullPointer";
    BinaryTokens[BinaryTokens["Hole"] = 1] = "Hole";
    BinaryTokens[BinaryTokens["Null"] = 2] = "Null";
    BinaryTokens[BinaryTokens["PBigInt"] = 3] = "PBigInt";
    BinaryTokens[BinaryTokens["NBigInt"] = 4] = "NBigInt";
    BinaryTokens[BinaryTokens["Boolean"] = 5] = "Boolean";
    BinaryTokens[BinaryTokens["String"] = 6] = "String";
    BinaryTokens[BinaryTokens["Undefined"] = 7] = "Undefined";
    BinaryTokens[BinaryTokens["UnsignedByte"] = 8] = "UnsignedByte";
    BinaryTokens[BinaryTokens["SignedByte"] = 9] = "SignedByte";
    BinaryTokens[BinaryTokens["UnsignedInt32"] = 10] = "UnsignedInt32";
    BinaryTokens[BinaryTokens["SignedInt32"] = 11] = "SignedInt32";
    BinaryTokens[BinaryTokens["UnsignedFloat64"] = 12] = "UnsignedFloat64";
    BinaryTokens[BinaryTokens["SignedFloat64"] = 13] = "SignedFloat64";
    BinaryTokens[BinaryTokens["Array"] = 14] = "Array";
    BinaryTokens[BinaryTokens["EmptyArray"] = 15] = "EmptyArray";
    BinaryTokens[BinaryTokens["ObjectReference"] = 16] = "ObjectReference";
    BinaryTokens[BinaryTokens["Date"] = 17] = "Date";
    BinaryTokens[BinaryTokens["BooleanObject"] = 18] = "BooleanObject";
    BinaryTokens[BinaryTokens["NumberObject"] = 19] = "NumberObject";
    BinaryTokens[BinaryTokens["StringObject"] = 20] = "StringObject";
    BinaryTokens[BinaryTokens["EmptyObject"] = 21] = "EmptyObject";
    BinaryTokens[BinaryTokens["Object"] = 22] = "Object";
    BinaryTokens[BinaryTokens["RegExp"] = 23] = "RegExp";
    BinaryTokens[BinaryTokens["Map"] = 24] = "Map";
    BinaryTokens[BinaryTokens["EmptyMap"] = 25] = "EmptyMap";
    BinaryTokens[BinaryTokens["WeakMap"] = 26] = "WeakMap";
    BinaryTokens[BinaryTokens["Set"] = 27] = "Set";
    BinaryTokens[BinaryTokens["EmptySet"] = 28] = "EmptySet";
    BinaryTokens[BinaryTokens["WeakSet"] = 29] = "WeakSet";
    BinaryTokens[BinaryTokens["ArrayBuffer"] = 30] = "ArrayBuffer";
    BinaryTokens[BinaryTokens["Int8Array"] = 31] = "Int8Array";
    BinaryTokens[BinaryTokens["Uint8Array"] = 32] = "Uint8Array";
    BinaryTokens[BinaryTokens["Uint8ClampedArray"] = 33] = "Uint8ClampedArray";
    BinaryTokens[BinaryTokens["Int16Array"] = 34] = "Int16Array";
    BinaryTokens[BinaryTokens["Uint16Array"] = 35] = "Uint16Array";
    BinaryTokens[BinaryTokens["Int32Array"] = 36] = "Int32Array";
    BinaryTokens[BinaryTokens["Uint32Array"] = 37] = "Uint32Array";
    BinaryTokens[BinaryTokens["Float32Array"] = 38] = "Float32Array";
    BinaryTokens[BinaryTokens["Float64Array"] = 39] = "Float64Array";
    BinaryTokens[BinaryTokens["DataView"] = 40] = "DataView";
})(BinaryTokens = exports.BinaryTokens || (exports.BinaryTokens = {}));
var BinaryPrimitives;
(function (BinaryPrimitives) {
    BinaryPrimitives["BigInt"] = "bigint";
    BinaryPrimitives["Boolean"] = "boolean";
    BinaryPrimitives["Number"] = "number";
    BinaryPrimitives["Object"] = "object";
    BinaryPrimitives["String"] = "string";
    BinaryPrimitives["Undefined"] = "undefined";
})(BinaryPrimitives = exports.BinaryPrimitives || (exports.BinaryPrimitives = {}));
exports.TypedArray = Object.getPrototypeOf(Int8Array);
